### PR TITLE
feat(indexer) processor module with basic l1Processor & l2Processor

### PR DESCRIPTION
**Pull request by hamdiallam**
Introduce a `processor` modules that exposes `L1Processor` and `L2Processors`. These structs define the core logic that indexes new batches retrieved from the fetcher. All database operations for a given batch are wrapped in a database transaction.

The L1 & L2 processors simply index block headers right now.

## Aside
- `indexer-refresh` command was updated such that the new indexer can now be run e2e.
- `EthClient#FinalizedBlockHeight` switched to "latest" blocktag instead of "finalized" blocktag to avoid dealing with issues with the local devnet. We'll address this
- L1Processor needs some additional config such that it doesn't start from genesis. It should start from the L1 block corresponding to L2 genesis

### CODE REVIEW

File: indexer/database/contract_events.go [hunk 1]
```diff
@@ -29,8 +29,6 @@ type L2ContractEvent struct {
 }
 
 type ContractEventsView interface {
-	L1ContractEventByGUID(string) (*L1ContractEvent, error)
```
Comment (hamdiallam) [inline indexer/database/contract_events.go:32]: Removing these since there's new use for them. Would rather add as-needed

File: indexer/database/blocks.go [hunk 2]
```diff
@@ -111,14 +108,19 @@ func (db *blocksDB) StoreLegacyStateBatch(stateBatch *LegacyStateBatch) error {
...
 
-func (db *blocksDB) FinalizedL1BlockHeight() (*big.Int, error) {
+// FinalizedL1BlockHeader returns the latest L1 block header stored in the database, nil otherwise
+func (db *blocksDB) FinalizedL1BlockHeader() (*L1BlockHeader, error) {
 	var l1Header L1BlockHeader
 	result := db.gorm.Order("number DESC").Take(&l1Header)
 	if result.Error != nil {
+		if errors.Is(result.Error, gorm.ErrRecordNotFound) {
+			return nil, nil
+		}
```
Comment (roninjin10) [inline indexer/database/blocks.go:118]: Why is this necessary?
```suggestion
```

File: indexer/processor/processor.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+type processor struct {
+	fetcher *node.Fetcher
+
+	db         *database.DB
+	processFn  processFn
+	processLog log.Logger
+}
+
+// Start kicks off the processing loop. This is a blocking operation and should be run within its own goroutine
+func (p processor) Start() {
```
Comment (roninjin10) [inline indexer/processor/processor.go:28]: What does op-node use for this?

**Review by roninjin10 — COMMENTED** [review]
need to refresh on gorm later today to give proper review

**Review by roninjin10 — COMMENTED** [review]
need to refresh on gorm later today to give proper review

File: indexer/database/blocks.go (hunk 2 above)
Comment (hamdiallam) [reply indexer/database/blocks.go:118]: Since we are returning pointers, I think it makes sense to have `nil` represent an non-existent entry when `err == nil`. Useful for the processor when determining if it should start from genesis or a specified height (last indexed block).

Gorm treats non existent entries as errors

File: indexer/processor/processor.go (hunk 3 above)
Comment (hamdiallam) [reply indexer/processor/processor.go:28]: Good question, not entirely sure cc @tynes

**Comment by hamdiallam** [comment]
DX-80

File: indexer/processor/processor.go (hunk 3 above)
Comment (hamdiallam) [reply indexer/processor/processor.go:28]: Looked into op-node. op-node/ defines a driver for walking the L1 chain. It's opinionated and specifically build for the op-node use case. This is simple enough for us to define here
